### PR TITLE
add dead MemRef elimination pass

### PR DESCRIPTION
**Pull request by adstraw**

### CODE REVIEW

**Review by jbruestle — CHANGES_REQUESTED** [review]
Very simple and clean, nice. Probably should add a negative test (i.e. you don't remove if there is a read op).

File: pmlc/dialect/affinex/transforms/dead_memref_elimination.cc [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+namespace pmlc::dialect::affinex {
+
+struct AffinexDeadMemRefElimination
+    : public AffinexDeadMemRefEliminationBase<AffinexDeadMemRefElimination> {
+
+  void runOnFunction() override {
+    llvm::SmallVector<Operation *, 8> opsToErase;
+    getFunction().walk([&](AllocOp alloc) {
+      auto memref = alloc.getResult();
+      for (auto user : memref.getUsers()) {
```
Comment (flaub) [inline pmlc/dialect/affinex/transforms/dead_memref_elimination.cc:17]: This should be `auto *` or even better: `Operation*`

File: pmlc/dialect/affinex/transforms/dead_memref_elimination.cc [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      }
+      opsToErase.push_back(alloc.getOperation());
+    });
+    for (auto *op : opsToErase) {
+      op->erase();
+    }
+  }
+};
+
+std::unique_ptr<mlir::Pass> createAffinexDeadMemRefElimination() {
```
Comment (flaub) [inline pmlc/dialect/affinex/transforms/dead_memref_elimination.cc:33]: nit: drop the `mlir::`

File: pmlc/dialect/affinex/transforms/dead_memref_elimination.cc (hunk 1 above)
Comment (adstraw) [reply pmlc/dialect/affinex/transforms/dead_memref_elimination.cc]: Done

File: pmlc/dialect/affinex/transforms/dead_memref_elimination.cc (hunk 2 above)
Comment (adstraw) [reply pmlc/dialect/affinex/transforms/dead_memref_elimination.cc]: Done, fixed this in AffinexMemRefDataFlow as well.

**Review by jbruestle — APPROVED** [review]

**Pull request merged**